### ISSUE TITLE
Cannot run xpra using Xdummy in parallel with desktop X environment

### ISSUE THREAD

**Issue by calvertvl** [NixOS/nixpkgs#24652]
## Issue description

My use case:
I have a desktop X environment that I am happy with (slim autologin + xmonad). I am also on remote systems (e.g. laptop), and want to be able to share certain program instances between the two at times. For example, a web browser or LibreOffice session, graphical IRC client, etc. - basically anything that does not use 3D acceleration or video playback.

I want to be able to SSH into the server or open a terminal, and do the following:
```
$ xpra start :10
$ DISPLAY=:10 ${CMD} & disown
```
This sets up an application that I'm interested in using.

I can then connect to the xpra server to use the applications via one of the following commands:
```
$ xpra attach ssh:<hostname>:10 # remote access
$ xpra attach :10 # local access
```

Yes, it would be nice to auto-start some of this, but that's relatively straightforward to  implement.

The problem:
* I can add `xpra` to `environment.systemPackages`; this installs fine.
* If the server command is overridden to use the Xvfb variant provided in the config file, this works.
* It **does not** work on NixOS with `xpra_Xdummy`. The logs indicate that it cannot find the `dummy` driver.

@offlinehacker mentioned the xpra.nix module in this context on #24644. This functionality is something I hadn't noticed before, but If I'm following correctly, I can't have a machine host both a desktop X environment and xpra while using the Xorg server.

Specifically, xpra is set up as a displayManager, and if I understand correctly, only one displayManager can be enabled on a NixOS system.

There are actually multple use cases for xpra: It is effectively a combination of display manager and window manager, in that it can provide forwarding for entire desktops (display manager case) or for applications (window manager case). 

### Steps to reproduce
I have `services.xserver.displayManager.slim.enable=true` in my configuration.nix.

If I add `services.xserver.displayManager.xpra.enable=true` without disabling slim, I get the following error on rebuild attempts:
```
error: The unique option `services.xserver.displayManager.job.execCmd' is defined multiple times, in `.../nixpkgs/nixos/modules/services/x11/display-managers/xpra.nix' and `.../nixpkgs/nixos/modules/services/x11/display-managers/slim.nix'.
(use ‘--show-trace’ to show detailed location information)
```

Disabling slim is not an acceptable solution as I need the desktop X environment...

## Technical details

* System: 17.09.git.6b5e5de (Hummingbird)
* Nix version: nix-env (Nix) 1.11.8
* Nixpkgs version: "17.09.git.6b5e5de0ef"

**Comment by offlinehacker** [comment]
New module will have to be made for that use case. something like
`service.xpra.enable`, these are two different use cases. One is display
manager and other is xpra that attaches to existing display manager. While
this is not really use case I have, i can help with development of such
module. Driver cannot be found for xdummy, because xorg config with path to
driver has to be passed to xdummy. In most other distros it's somewhere
under /etc.

On Wed, 5 Apr 2017, 20:01 Victor Calvert, <notifications@github.com> wrote:

Issue description

My use case:
I have a desktop X environment that I am happy with (slim autologin +
xmonad). I am also on remote systems (e.g. laptop), and want to be able to
share certain program instances between the two at times. For example, a
web browser or LibreOffice session, graphical IRC client, etc. - basically
anything that does not use 3D acceleration or video playback.

I want to be able to SSH into the server or open a terminal, and do the
following:

$ xpra start :10
$ DISPLAY=:10 ${CMD} & disown

This sets up an application that I'm interested in using.

I can then connect to the xpra server to use the applications via one of
the following commands:

$ xpra attach ssh:<hostname>:10 # remote access
$ xpra attach :10 # local access

Yes, it would be nice to auto-start some of this, but that's relatively
straightforward to implement.

The problem:

   - I can add xpra to environment.systemPackages; this installs fine.
   - If the server command is overridden to use the Xvfb variant provided
   in the config file, this works.
   - It *does not* work on NixOS with xpra_Xdummy. The logs indicate that
   it cannot find the dummy driver.

@offlinehacker <https://github.com/offlinehacker> mentioned the xpra.nix
module in this context on #24644
<https://github.com/NixOS/nixpkgs/pull/24644>. This functionality is
something I hadn't noticed before, but If I'm following correctly, I can't
have a machine host both a desktop X environment and xpra while using the
Xorg server.

Specifically, xpra is set up as a displayManager, and if I understand
correctly, only one displayManager can be enabled on a NixOS system.

There are actually multple use cases for xpra: It is effectively a
combination of display manager and window manager, in that it can provide
forwarding for entire desktops (display manager case) or for applications
(window manager case).
Steps to reproduce

I have services.xserver.displayManager.slim.enable=true in my
configuration.nix.

If I add services.xserver.displayManager.xpra.enable=true without disabling
slim, I get the following error on rebuild attempts:

error: The unique option `services.xserver.displayManager.job.execCmd'
is defined multiple times, in
`.../nixpkgs/nixos/modules/services/x11/display-managers/xpra.nix' and
`.../nixpkgs/nixos/modules/services/x11/display-managers/slim.nix'.
(use ‘--show-trace’ to show detailed location information)

Disabling slim is not an acceptable solution as I need the desktop X
environment...
Technical details

   - System: 17.09.git.6b5e5de (Hummingbird)
   - Nix version: nix-env (Nix) 1.11.8
   - Nixpkgs version: "17.09.git.6b5e5de0ef"

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/NixOS/nixpkgs/issues/24652>, or mute the thread

**Comment by jb55** [comment]
> New module will have to be made for that use case. something like
> `service.xpra.enable`, these are two different use cases. One is display
> manager and other is xpra that attaches to existing display manager. 

The latter is my usecase as well. I was trying to figure out how to use
xpra in nixos with an existing display manager and I couldn't get it
working, I ran into the same issues.

👍 for a `serivce.xpra`. Thanks if you get around to adding this!

**Comment by calvertvl** [comment]
I'll try to sort out a service, probably in a similar manner to either the emacs daemon or xinetd.

Here's what I'm thinking right now:
* xpra needs $DISPLAY for each X server process
* The most flexible approach is to do this similar to xinetd, specifying a list of values for $DISPLAY and then the appropriate configuration for each (eventually including allowing access to more than one user, as that is a possible use case).
* It also makes sense to provide a simple "service.xpra.enable" or similar that defaults to DISPLAY=:$UID or something sensible (auto-setup one process per user)

Note that all instances would share the same xpra & xorg derivations: the only real difference are things that are set at runtime through the xpra config.

**Comment by offlinehacker** [comment]
Btw one unrelated question. What is latency with running applications remotely? I'm running xpra in vms, where I also enabled mmap(so it's zero copy) and still experience some minor latency, that sometimes is annoying if I code fast or want to browse web smoothly.

**Comment by ryanorendorff** [comment]
Does a service like this assist in this process?

https://github.com/Nekroze/xpra.nix/blob/master/test.nix

I'm pretty new to Nix so I am not quite how to get the service to run.

**Comment by totaam** [comment]
> I'm running xpra in vms, where I also enabled mmap(so it's zero copy) and still experience some minor latency

mmap is not possible with vms.
Try vsock / ivshmem to speed things up.

Most latency issues should have been resolved since xpra 2.5.x

**Comment by samueldr** [comment]
Closing, as with #87985, it is assumed this issue should be fixed.

If *this particular issue* is not fixed, feel free to reopen, or ask to reopen. For another problem, open a new issue, and CC the relevant people working on better Xpra support (see #86164).

Thanks!
